ASoC: SOF: ipc-topology: Add float and 8-bit formats

singalsu opened the pull request:
See commit messages for the changes.

singalsu committed:
ASoC: SOF: ipc4-topology: Add support for 8-bit formats

This patch enables use of 8-bit unsigned, A-law, and mu-law
sample formats with IPC4 SOF.

The ipc4-topology.h is updated with IPC4 sample types. The
purpose is to to convert ALSA types to IPC4 types.

The functions of_ipc4_update_hw_params(), sof_ipc4_get_valid_bits(),
and new function sof_ipc4_get_sample_type() are updated to handle
the sample type conversions.

The function sof_ipc4_prepare_copier_module() is updated to set
the DMA SCS bit for all non 32 bits sample types.

The change to function sof_ipc4_get_valid_bits() returns 8 bits
for these ALSA formats.

The change to function sof_ipc4_prepare_copier_module() is
needed to handle properly all non 32-bit formats with SCS
bit set.

To support playback with new 8 bits types, the
sof_ipc4_init_input_audio_fmt() function is updated to get
the sample type and use it in search for copier input pin
format.

To support capture, the sof_ipc4_init_output_audio_fmt()
is updated similarly. Since the function uses separate
out_ref_type argument, instead of single parameters struct,
the out_ref_type needs to be added to every user of the
function. Therefore functions sof_ipc4_prepare_copier_module(),
sof_ipc4_prepare_gain_module(), sof_ipc4_prepare_mixer_module(),
sof_ipc4_prepare_src_module(), and sof_ipc4_prepare_process_module()
are updated to set the out_ref_type.

Signed-off-by: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>

singalsu commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1442,6 +1442,25 @@ static int sof_ipc4_get_valid_bits(struct snd_sof_dev *sdev, struct snd_pcm_hw_p
+	case SNDRV_PCM_FORMAT_S16_LE:
+	case SNDRV_PCM_FORMAT_S24_LE:
+	case SNDRV_PCM_FORMAT_S32_LE:
+		return SOF_IPC4_TYPE_SIGNED_INTEGER;
```
Normal formats fail. This seems to need to be SOF_IPC4_TYPE_LSB_INTEGER. And I think naming like in topology SOF_IPC4_SAMPLE_TYPE_x would be better.

singalsu replied:
Still not right, playback without conversion fails.

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1483,8 +1508,9 @@ static int sof_ipc4_init_input_audio_fmt(struct snd_sof_dev *sdev,
+		type = SOF_IPC4_AUDIO_FORMAT_CFG_SAMPLE_TYPE(fmt->fmt_cfg);
 		if (params_rate(params) == rate && params_channels(params) == channels &&
-		    sample_valid_bits == valid_bits)
+		    sample_valid_bits == valid_bits && sample_type == type)
```
@singalsu do you have a corresponding topology change for this? Dont we default to LSB_INTEGER in topology?

singalsu replied:
Yes, the topology that I use for testing is in https://github.com/thesofproject/sof/pull/9980 .

The MSB integer and LSB integer usage is confusing. I would use in topology signed integer for s16 and s32. At least that's what comes to my mind first. Though they would be also valid MSB integer and LSB integer since the valid size is the same as container type. The ALSA S24_4LE as low part aligned would be LSB integer.

I need this check to handle more special types to separate A-law from integer 8-bit (that is also special as unsigned) . Also float (we miss support)  is 32 bits valid and container size and need be set to IPC4 float type.

singalsu replied:
While in firmware comment it looks like the types indicate endiannes:

```
enum ipc4_sample_type {
	IPC4_TYPE_MSB_INTEGER = 0,	/**< integer with Most Significant Byte first */
	IPC4_TYPE_LSB_INTEGER = 1,	/**< integer with Least Significant Byte first */
	IPC4_TYPE_SIGNED_INTEGER = 2,
	IPC4_TYPE_UNSIGNED_INTEGER = 3,
	IPC4_TYPE_FLOAT = 4,
	IPC4_TYPE_A_LAW = 5,
	IPC4_TYPE_MU_LAW = 6,
};

```

singalsu replied:
@ranj063 The playback direction works now, the issue was in topology. Yes, the topologies default to LSB_INTEGER and I've not changed that. I decided to use IPC4_TYPE_UNSIGNED_INTEGER for ALSA U8 since there is also ALSA S8. If need support for that it would be (maybe best as) IPC4_TYPE_SIGNED_INTEGER. Since it's a byte, the IPC4_TYPE_LSB_INTEGER would be confusing. But so far I'm not aware of any wav files with S8 format, so I'm not planning to add it now.

I currently work with capture direction. With the latest PR here, I can run capture with same formats but I haven't checked that it's correct yet.

ranj063 replied:
@singalsu I think we need to test this patch alone (as a separate PR) with the current topologies to make sure we pass isnt it? We've never included the sample_type in the tests before and we should make sure that it is correct in existing topologies before we start adding support for 8-bit formats

singalsu replied:
Yep, in tests below seems nocodec fails ([  264.082122] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: Seeking for type -949225600). I need to fix those uninitialized types first in topologies.

singalsu replied:
Actually above was a bug in this patch with use of unset type. New try soon with this kernel patches CI.

singalsu commented:
SOFCI TEST

singalsu commented:
I combined all the patches, not sure if this can be split to preparation and feature enable.

I've tested this with script like below UPX i11 device with this kernel patch and firmware https://github.com/thesofproject/sof/pull/9980. Then copied from DUT the nocodec loopback wav files and checked visually them with spectrogram in audio editor. The chirp should sound and look clean with all formats combinations.

singalsu commented:
No changes, just rebased and improved commit text to re-launch SOFCI.

ujfalusi commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1443,14 +1460,15 @@ static int sof_ipc4_init_output_audio_fmt(struct snd_sof_dev *sdev,
 		_out_rate = fmt->sampling_frequency;
 		_out_channels = SOF_IPC4_AUDIO_FORMAT_CFG_CHANNELS_COUNT(fmt->fmt_cfg);
 		_out_valid_bits = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(fmt->fmt_cfg);
+		_out_type = SOF_IPC4_AUDIO_FORMAT_CFG_SAMPLE_TYPE(fmt->fmt_cfg);
```
it is **granted** that the SOF_IPC4_AUDIO_FORMAT_CFG_SAMPLE_TYPE() for non 8 bit formats are _always_ matching, right?
When used with older topologies where the type is not set.

singalsu replied:
It looks like the sample type was first set in this December 2021 patch. The error would be with sample type not set.

```
commit 663174320904c0bc045f2d60cbe62035147c9736 (HEAD)
Author: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>
Date:   Sat Dec 11 21:03:50 2021 -0800

    topology2: common: Add new class definition for audio_format
    
    Add definition for audio format class for defining the input and output
    audio formats for each components in cavs topology.
    
    Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

```

It looks like the risk of encountering such topology is very low, only own tplg builds can be such. SOF v2.7 was first release for IPC4 and it was September 2023.

ujfalusi commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1984,10 +2032,20 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 				pipeline->msg.primary |= SOF_IPC4_GLB_CHAIN_DMA_SCS_MASK;
 
+			/* Set SCS bit for 8 and 16 bit formats */
+			switch (params_format(fe_params)) {
```
```c
if (params_physical_width(fe_params) <= 16)
	pipeline->msg.primary |= SOF_IPC4_GLB_CHAIN_DMA_SCS_MASK;
```

singalsu replied:
Yep, much better that way.

ujfalusi commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1480,6 +1502,25 @@ static int sof_ipc4_get_valid_bits(struct snd_sof_dev *sdev, struct snd_pcm_hw_p
+		return SOF_IPC4_TYPE_LSB_INTEGER;
+	default:
+		dev_err(sdev->dev, "invalid pcm sample type %d\n", params_format(params));
+		return -EINVAL;
```
HDMI supports IEC958_SUBFRAME_LE, that should be not an error.

singalsu replied:
I guess I'll add  `case SNDRV_PCM_FORMAT_IEC958_SUBFRAME_LE:` to return `SOF_IPC4_TYPE_LSB_INTEGER` though it isn't exactly right. I need to get his somehow tested, maybe with my TV and HDMI.

singalsu replied:
@ujfalusi This seems to be called only for capture. I was never able to get SNDRV_PCM_FORMAT_IEC958_SUBFRAME_LE, since HDMI is only for playback. What do you think, should I remove the added case since it doesn't get executed?

As alternative way, this function could return by default SOF_IPC4_TYPE_LSB_INTEGER and only warn if ALSA PCM format is unknown. I don't know really what to do other than this conversion from ALSA to IPC4 needs to be done for 8-bit and next to-be-added float sample type.

singalsu committed:
ASoC: SOF: ipc4-topology: Add support for float sample type

This patch adds support for FLOAT_LE ALSA type for playback and
capture for IPC4 systems. Functions sof_ipc4_update_hw_params(),
sof_ipc4_get_valid_bits() and sof_ipc4_get_sample_type()
are updated to handle SOF_IPC4_TYPE_FLOAT and
SNDRV_PCM_FORMAT_FLOAT_LE.

Signed-off-by: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>

ujfalusi reviewed (approved):
@singalsu, looks good to me!
